src/pagination.rs

```diff
@@ -242,6 +247,142 @@ fn sort_posts(posts: &mut Vec<&liquid::value::Value>, config: &PaginationConfig)
+        Category {
+            cat_name: name.to_vec(),
+            posts: vec![],
+            sub_cats: vec![],
+        }
+    }
+
+    fn add_post(&mut self, post: &'a liquid::value::Value) {
+        self.posts.push(&post)
+    }
+}
+
+// only used in parse_categories_list, to transform liquid::value::Array to type with Ord trait
+fn liquid_array_to_string(arr: &liquid::value::Array) -> String {
+    let mut s: String = arr
+        .iter()
+        .map(|i| {
+            let mut s = i.to_str().to_string();
+            s.push('/');
+            s
+        })
+        .collect();
+    s.pop(); // remove last '/'
+    s
+}
+
+// construct a hierarchy of Categories with their posts from a list of categories
+fn parse_categories_list<'a, 'b>(
+    parent: &'b mut Category<'a>,
+    cur_idx: usize,
+    categories: &[liquid::value::Value],
+    post: &'a liquid::value::Value,
+) -> Result<()> {
+    if cur_idx <= categories.len() {
+        let cat_full_name = categories[..cur_idx].to_vec();
+        let mut cur_cat = if let Ok(idx) = parent
+            .sub_cats
+            .binary_search_by_key(&liquid_array_to_string(&cat_full_name), |c| {
+                liquid_array_to_string(&c.cat_name)
+            }) {
+            &mut parent.sub_cats[idx]
+        } else {
+            let last_idx = parent.sub_cats.len();
+            parent.sub_cats.push(Category::new(&cat_full_name));
+            // need to sort for binary_search_by_key
+            // need to convert into String to get Ord trait
+            parent
+                .sub_cats
+                .sort_by_key(|c| liquid_array_to_string(&c.cat_name));
+
+            &mut parent.sub_cats[last_idx]
+        };
+
+        if cur_idx == categories.len() {
+            cur_cat.add_post(&post);
+        } else {
+            parse_categories_list(&mut cur_cat, cur_idx + 1, &categories, &post)?;
+        }
+    }
+    Ok(())
+}
+
+fn distribute_posts_by_categories<'a>(
+    all_posts: &[&'a liquid::value::Value],
+) -> Result<Category<'a>> {
+    let mut root = Category::new(&vec![]);
+    for post in all_posts {
+        if let Some(categories) = extract_categories(post) {
+            parse_categories_list(&mut root, 1, categories.as_slice(), &post)?;
+        }
+    }
+    Ok(root)
+}
+
+// walk the categories tree and construct Paginator for each node,
+// filling `pages` and `indexes` accordingly
+fn walk_categories<'a, 'b>(
+    category: &'b mut Category<'a>,
+    config: &PaginationConfig,
+    doc: &Document,
+) -> Result<Vec<Paginator>> {
+    let mut cur_cat_paginators_holder: Vec<Paginator> = vec![];
+    if !category.cat_name.is_empty() {
+        sort_posts(&mut category.posts, &config);
+        let cur_cat_paginators = create_all_paginators(
+            &category.posts,
+            doc,
+            &config,
+            Some(&liquid::value::Value::array(category.cat_name.clone())),
+        )?;
+        if !cur_cat_paginators.is_empty() {
```

Comment:
Under what condition would a category exist but have no paginators?

Or does this mean that all of the items for the category fit on one page?  Can't quite remember how this was set up.

Reply:
Imagine a post in `/cat1/cat2`, `cat1` has no post so after `create_all_paginators`, we have none. Hence creation of a Paginator with no post.

Reply:
Ah, yes.